pkg/webui/console/store/middleware/logics/user-preferences.js

```diff
@@ -18,16 +18,159 @@ import createRequestLogic from '@ttn-lw/lib/store/logics/create-request-logic'
 
 import * as userPreferences from '@console/store/actions/user-preferences'
 
+const addBookmarkToResult = (result, entity, element) => {
+  if (!(entity in result.perEntityBookmarks)) {
+    result.perEntityBookmarks[entity] = []
+    result.perEntityBookmarks[entity].push(element)
+  } else {
+    result.perEntityBookmarks[entity].push(element)
+  }
+  if (!(entity in result.totalCount.perEntityTotalCount)) {
+    result.totalCount.perEntityTotalCount[entity] = 1
+  } else {
+    result.totalCount.perEntityTotalCount[entity] += 1
+  }
+
+  return result
+}
+
+const getBookmarksThroughPagination = async userId => {
+  let page = 1
+  const limit = 1000
+  let totalCount = Infinity
+  let result = {
+    bookmarks: [],
+    perEntityBookmarks: {},
+    totalCount: {
+      totalCount: 0,
+      perEntityTotalCount: {},
+    },
+  }
+
+  while ((page - 1) * limit < totalCount) {
+    // Get the next page of bookmarks.
+    // eslint-disable-next-line no-await-in-loop
+    const response = await tts.Users.getBookmarks(userId, { page, limit, order: '-created_at' })
+    response.bookmarks.forEach(element => {
+      const entityIds = element.entity_ids
+      const entity = Object.keys(entityIds)[0].replace('_ids', '')
+      result = addBookmarkToResult(result, entity, element)
+    })
+
+    result = {
+      bookmarks: [...result.bookmarks, ...response.bookmarks],
+      perEntityBookmarks: { ...result.perEntityBookmarks },
+      totalCount: {
+        ...result.totalCount,
+        totalCount: response.totalCount,
+      },
+    }
+
+    totalCount = response.totalCount
+    page += 1
+  }
+
+  return result
+}
+
+const getBookmarksPerEntityThroughPagination = async (
+  userId,
+  requestedEntity,
+  start,
+  end,
+  limit,
+) => {
+  let page = 1
+  let totalCount = limit
+  const result = {
+    perEntityBookmarks: {
+      [requestedEntity]: [],
+    },
+    perEntityTotalCount: {
+      [requestedEntity]: 0,
+    },
+  }
+
+  while (result.perEntityBookmarks[requestedEntity].length < totalCount) {
+    // eslint-disable-next-line no-await-in-loop
+    const response = await tts.Users.getBookmarks(userId, { page, limit, order: '-created_at' })
```

Comment:
We should also consider fetching the additional data from the individual entity endpoint here and decorate the result with them, instead of doing that separately.

Reply:
But let's not touch this further now as @nicholaspcr is already working on entity filters in the backend, so we will need to refactor this soon anyway.

Reply:
true, link to the [PR](https://github.com/TheThingsNetwork/lorawan-stack/pull/7041) as to make it easy for tracking.